### ISSUE TITLE
Crash from runtime error inside a withState

### ISSUE THREAD

**Issue by duvenaud** [google-research/dex-lang#231]
Thanks to @danieldjohnson 's help I'm making good progress on the simplex demo.  But I hit a crash when trying to read out the final solution, which involves fancy indexing and variant matching inside a ```withState```.

Edit: Ignore this example - @danieldjohnson wrote a much smaller MWE below.

```
def ColIx (cons:Type) (vars:Type) : Type =
  {variables:vars | slacks:cons}

def build_tableau (_:Eq cons) ?=>
    (ineqs: cons=>{coeffs:vars=>Float})
     : cons=>(ColIx cons vars) =
  for r:cons. {|slacks=r|}

fake_constraints = for i:(Fin 5). {coeffs=for j:(Fin 2). 1.0}
basics = build_tableau fake_constraints

def extractSolution (basics:cons=>(ColIx cons vars)) : vars=>Float =
  snd $ withState zero \varRef.
    for v.   -- in vars.
      curvar = basics.v  -- get column index
      case curvar of  -- match only on 'variables', not slacks.
        {|variables=curvar|} -> () -- would write to varRef here if it didn't crash
  

:p extractSolution basics  -- This line causes the crash
```

```
Not an imp atom, or mismatched dest: Dest (), and (%anyVal Unit)
CallStack (from HasCallStack):
  error, called at src/lib/Imp.hs:654:19 in dex-0.1.0.0-IYxyzQADM8eGhInlFxB9Qk:Imp
```

To help understand what the code is doing in plain English: I have an array of column indices.  The column indices come in two flavors: "variables" and "slacks".  I'm trying to construct an array indexed by variables, which will have value 0 if there is no such variable indexed in ```basics```, but otherwise will index into the table.

**Comment by danieldjohnson** [comment]
This seems like a bug in missing-case handling. I think `%anyVal t` is the value returned by an unwritten case alternative. But that's supposed to turn into a runtime error, not a compile-time error.

`basics.v` is of type `{variables:vars | slacks:cons}`, so it could be either a variable or a slack. I think Dex is interpreting what you wrote as "if its a variable, do this, otherwise throw a runtime error" and then there's a compiler bug when trying to compile the runtime error. Did you mean to only handle the case where it is a variable? Maybe this would work:
```
      case basics.v of
        {|variables=curvar|} -> () -- todo: write to varRef here
        {|slacks=_|} -> () -- do nothing
```

**Comment by danieldjohnson** [comment]
Smaller repro of the same runtime error issue:
```
x : Unit => Unit = fst $ withState () $ \_. (for i. todo)
```

**Comment by duvenaud** [comment]
Wow, thanks for the quick reply.  Yes, your suggestion of handling both cases does resolve the problem, thanks!  I had tried that at one point without seeing success, but I must still have had another problem.

Also thanks for finding a simpler MWE.

**Comment by apaszke** [comment]
This is fixed now.
